### PR TITLE
[`ruff`] Unformatted special comments (`RUF037`)

### LINKED ISSUE #10160: Request: Enforce space after colon (`:`) for special comments
I've open the same request over at https://github.com/PyCQA/flake8-pyi/issues/465 , but I think that Ruff could generalize this, or at least expand on supported keywords. 

`# pyright:ignore` --> `# pyright: ignore`
`# type:ignore` --> `# type: ignore`
`# noqa:Y011` --> `# noqa: Y011`
`# fmt:on` -> `# fmt: on`
`# ruff:noqa` --> `# ruff: noqa`
`# ruff:noqa:F811` --> `# ruff: noqa: F811` (or at least `# ruff: noqa:F811` if generalized)
`# flake8:noqa:F811` --> `# flake8: noqa: F811` (or at least `# flake8: noqa:F811` if generalized)
`# ruff:isort:skip_file` -> `# ruff: isort: skip_file` (or at least `# ruff: isort:skip_file` if generalized)
`# isort:skip_file` -> `# isort: skip_file`
...

Basically the same as https://docs.astral.sh/ruff/rules/missing-space-after-todo-colon/ but with a lot more keywords

Formatters (black and Ruff) already take of adding a space after `#`

### PR DESCRIPTION
**Pull request by InSyncWithFoo**
## Summary

Resolves #10160.

## Test Plan

`cargo nextest run` and `cargo insta test`.

### CODE REVIEW

File: crates/ruff_linter/src/rules/ruff/rules/unformatted_special_comment.rs [hunk 1]
```diff
@@ -0,0 +1,278 @@
+use crate::Locator;
+use once_cell::sync::Lazy;
```
Comment (MichaReiser) [inline crates/ruff_linter/src/rules/ruff/rules/unformatted_special_comment.rs:2]: You can use https://doc.rust-lang.org/std/sync/struct.LazyLock.html

File: crates/ruff_linter/src/checkers/noqa.rs [hunk 2]
```diff
@@ -51,8 +51,15 @@ pub(crate) fn check_noqa(
 
         match &exemption {
             FileExemption::All(_) => {
-                // If the file is exempted, ignore all diagnostics.
-                ignored_diagnostics.push(index);
+                // If the file is exempted, ignore all diagnostics,
+                // save for RUF104, which operates on `# noqa` comments
+                // and thus needs to be suppressed explicitly.
+                if !matches!(diagnostic.kind.rule(), Rule::UnformattedSpecialComment)
```
Comment (MichaReiser) [inline crates/ruff_linter/src/checkers/noqa.rs:57]: Can you explain the reasoning for this special handling? I think I would find this behavior surprising and I'm not sure if it justifies the added complexity

File: crates/ruff_linter/src/rules/ruff/mod.rs [hunk 3]
```diff
@@ -384,6 +384,10 @@ mod tests {
     }
 
     #[test_case(Rule::ZipInsteadOfPairwise, Path::new("RUF007.py"))]
+    #[test_case(Rule::UnformattedSpecialComment, Path::new("RUF104.py"))]
+    #[test_case(Rule::UnformattedSpecialComment, Path::new("RUF104_1.py"))]
+    #[test_case(Rule::UnformattedSpecialComment, Path::new("RUF104_2.py"))]
+    #[test_case(Rule::UnformattedSpecialComment, Path::new("RUF104_3.py"))]
```
Comment (MichaReiser) [inline crates/ruff_linter/src/rules/ruff/mod.rs:390]: Are these files testing different aspects? If so, can we give them more meaningful names than `_1` etc.?

File: crates/ruff_linter/src/rules/ruff/rules/unformatted_special_comment.rs [hunk 4]
```diff
@@ -0,0 +1,553 @@
...
+    /// `# fmt: on`, `# fmt: off`, `fmt: skip`
+    Fmt(String),
+    /// `# isort: skip_file`
+    Isort(String),
+    /// `# mypy: ignore-errors`
+    Mypy(String),
+    /// `# nopycln: import`
+    Nopycln(String),
+    /// `# pyright: strict`, `# pyright: ignore[reportFoo]`
+    Pyright(String),
```
Comment (MichaReiser) [inline crates/ruff_linter/src/rules/ruff/rules/unformatted_special_comment.rs:34]: I'm somewhat hesitant supporting non-ruff suppression comments:
1. All other `RUF1XX` rules are specific to ruff pragma comments
2. I don't see us as the authority defining how `pyright` and `mypy` comments should be formatted.

**Review by MichaReiser — COMMENTED** [review]
Thanks for working on this rule. 

I think we need to define the scope of this rule more carefully, especially when it comes to non-ruff suppression comments.

1. All other `RUF1XX` rules are specific to ruff pragma comments
2. I don't see us as the authority defining how `pyright` and `mypy` comments should be formatted.

I see two possible versions of this rule:

* One specific to ruff's suppression comments that enforces consistent formatting, including how rule codes are formatted
* A general purpose rule that handles all kinds of pragma comments. It only enforces a single space between `#` and the keyword and, optionally, that the colon is followed by a space. We can add a configuration option to support custom pragma codes. This is not a `RUF1XX` rule.

**Comment by MichaReiser** [comment]
@AlexWaygood do you have any recommendations on the rule's scope?

File: crates/ruff_linter/resources/test/fixtures/ruff/RUF104.py
Comment (DetachHead) [inline crates/ruff_linter/resources/test/fixtures/ruff/RUF104.py:1]: are `# pyright:` comments supported too?

**Comment by AlexWaygood** [comment]
I agree that we're getting into very hairy territory if we start formatting other tools' pragma comments by default. However, it does seem useful to have a rule that does generalised formatting of pragma comments.

I like @MichaReiser's second suggestion in https://github.com/astral-sh/ruff/pull/14111#pullrequestreview-2427244566 -- I would suggest a rule
- That is not in the `RUF1XX` category (which as @MichaReiser says is a category specifically for Ruff pragma comments)
- That by default, only formats Ruff pragma comments. (And maybe also `type: ignore` comments, since they are standardised across tools according to a specification?)
- That can be configured so that it also includes other tool-specific pragma comments such as `pyright: ignore`, `isort: skip_file` and `pragma: nocover`

**Comment by AlexWaygood** [comment]
> Currently the rule is handling (aside from `# type`) `# noqa`, `# ruff: noqa`, `# flake8: noqa`, `# isort`, `# ruff: isort` and `# fmt`. Aren't these all comments that Ruff reads?

They are! I wasn't necessarily criticising anything you were doing -- was just giving my response to Micha's request in https://github.com/astral-sh/ruff/pull/14111#issuecomment-2470227442 on what the scope of the rule should be :-)

> I'm in favor of limiting this rule to just Ruff comments, as I have already rewritten this PR a couple of times and I don't feel like redoing it all over again. I'll give custom comments their own rule in another PR.

Feel free to keep this PR small and not implement any configurability now. Small, incremental PRs are definitely encouraged! However, if you want to open a followup PR, please make this rule configurable in the followup PR rather than creating a separate rule just for custom comments. I think Micha and I are aligned that a single, configurable rule would be preferable to two rules that do very similar things.

**Comment by MichaReiser** [comment]
I do think that we can drastically simplify the implementation because we no longer need to special case different comments. All we need is to match for `# (fmt|noqa|...)` and then see if it has an optional `:` following it.

**Comment by MichaReiser** [comment]
The problem I see with special casing is that it isn't a good fit for a general-purpose, extensible rule; it should instead be a specific rule for `noqa`. 

The question is if we can come up with some general rules that work for all suppression comments without being specific to any pragma comment and that allows easy configuration:

* Require a space between `#` and `name` in # <name>`
* Require no space left of a `:` but require a space right of it: `# name: ignore` ok, `# name:ignore` not-ok
* Require no a space left of a `,` but require a space right of it: `# type: ignore[a, b]` is ok, `type: ignore[a,b]` isn't
* Require no space around `[` but require a space right of `]` if it isn't the end of the comment.

The rule doesn't require that any suppression comment is valid. E.g. it accepts `#type: ignore, a, b, cd` t

Note: What's interesting is that e.g. pyright's documentation uses a space before `[`: `# pyright: ignore [reportPrivateUsage, reportGeneralTypeIssues]` whereas mypy recommends using no-space `# type: ignore[<code>]` (which I prefer)

**Comment by Avasam** [comment]
> A general purpose rule that handles all kinds of pragma comments. It only enforces a single space between # and the keyword and, optionally, that the colon is followed by a space. We can add a configuration option to support custom pragma codes. 

Making this rule extensible is also what I've suggesting in https://github.com/astral-sh/ruff/issues/10160#issuecomment-1971452721 . Even if it comes as a follow-up PR, I'd like if it could handle comments that Ruff doesn't need to be aware of (like pyright)

**Comment by egberts** [comment]
Which is why I stop using RUFF. 

Due to Annoyance.

**Pull request closed without merging**